### ISSUE TITLE
Diameter .dia not automatically inheriting indirect inherits

### ISSUE THREAD

**Issue by pespin** [erlang/otp#8235]
**Describe the bug**

TS 29.273 SAR contains MIP6-Agent-Info AVP from RFC5447, which in turn contains MIP-Home-Agent-Address  AVP from RFC 4004.

In summary, specs A, B, C, where C depends on B, and B depends on A.

- Create a "diameter_3gpp_ts29_273_swx.dia" file containing SAR message definition and which inherits ("@inherits diameter_rfc5447", diameter_rfc5447.dia) to reference "MIP6-Agent-Info".
- Create a "diameter_rfc5447.dia" file containing MIP6-Agent-Info definition and which inherits ("@inherits diameter_rfc4004", diameter_rfc4004.dia) to reference "MIP-Home-Agent-Address".
- Create a "diameter_rfc4004.dia" file containing MIP-Home-Agent-Address definition.

Hence, diameter_3gpp_ts29_273_swx.dia is not directly inheriting RFC4004: no "@inherits diameter_rfc4004" in diameter_3gpp_ts29_273_swx.dia.

Then create some code which generates a SAR message containing MIP6-Agent-Info which in turn contains a MIP-Home-Agent-Address, and transmits it with diameter:call().
The diameter app will compile just fine, but it will fail to encode it, because the generated diameter_3gpp_ts29_273_swx.erl/hrl file lacks the code to encode MIP-Home-Agent-Address :
`[error] <0.164.0>@aaa_diameter_swx:handle_call:{175,13} Error: encode`

If one adds the explicit "@inherits diameter_rfc4004" in diameter_3gpp_ts29_273_swx.dia, then one can see that extra code is generated in diameter_3gpp_ts29_273_swx.erl/hrl and encoding works fine.

So, right now, one must inherit explicitly all indirect dependency .dia files in order to get a working encoder/decoder.

**To Reproduce**

The steps mentioned above can be seen in this branch "pespin/diameter-bug" I created in osmo-epdg.git:
https://gitea.osmocom.org/erlang/osmo-epdg/commits/branch/pespin/diameter-bug
One can see 3 commits in that branch:

- [Import generated code with bug](https://gitea.osmocom.org/erlang/osmo-epdg/commit/67555a7cc4f03da280098c6d2f4d381280e2f862): In there I simply import the generated code with the bug present (no explicit inherit of the indirect dependency rfc4004b in swx file).
- [diameter_3gpp_ts29_273_swx.dia: Fix missing dependency RFC4004](https://gitea.osmocom.org/erlang/osmo-epdg/commit/84b3ffafb0ed29d6c909209be6af203909657107): In here I apply the indirect dependency rfc4004 in diameter_3gpp_ts29_273_swx.dia.
- [Extra generated code after adding indirect dependency](https://gitea.osmocom.org/erlang/osmo-epdg/commit/b147f4f19aaade1634d9182853a6d1de31ee72cf): In here I commit the differences of the output .erl/.hrl after regenerating the code once the indirect dependency was applied.

By looking at the last commit one can see code which imho should have already been there already before inheriting the indirect spec.

**Expected behavior**

I see 2 solutions to the problem:
- Support automatically inheriting indirect dependencies. If there's spec "A, B and C;  C inherits B;  B inherits A", then when C inherits B, it should recursively generate code from A too.
- *Not* support automatically inheriting indirect dependencies, but at least fail at compile time saying there's a missing inherit/type, since clearly the generated code will fail at runtime.

**Affected versions**

I'm seeing this with erlang OTP 26.2.2-1 package from Archlinux.

**Additional context**

**Comment by bmk** [comment]
We are looking into this.

**Comment by Mikaka27** [comment]
Hi @pespin, could you test the draft PR and see whether it works for you?
https://github.com/erlang/otp/pull/10149

**Comment by pespin** [comment]
I tested again the same steps mentioned during ticket report above, this time on top of osmo-epdg master and with my Archlinux erlang otp 28.0.2-1, and I got same mentioned results as explained earlier (just to update the current "released" state since some time passed by since I reported the issue). This can be found in osmo-epdg.git branch "pespin/diameter-bug2".

I then cloned OTP from branch "michal/diameter/fix-inheriting-of-indirect-inherits/otp-19626" as requested, built it:
```
export ERL_TOP=`pwd`
./configure --prefix=/opt/erlang-otp
make
sudo make install
```

Then run again the same test in osmo-epdg pointing rebar3 to use erl from the compiled OTP (I verified with "ps -ef" the expected erl and libs are taken):
```
PATH="/opt/erlang-otp/bin/:$PATH" make
```

Result: Exact same bug behavior, see osmo-epdg.git branch "pespin/diameter-bug3-pr".

I also tested with a simplified a.dia b.dia c.dia provided by @lynxis and it also sames same behavior in both OTP 28.0.2- and the one from your PR.

**Comment by pespin** [comment]
@Mikaka27 My bad I forgot to add the "indirect_inherits" during the test, I first though the new behavior would be applied by default. testing again...

**Comment by pespin** [comment]
@Mikaka27 took me a while to fix our .dia dependencies in osmo-epdg when enabling indirect_inherits ;)

I confirm now code is generated even if removing "@inherits diameter_rfc4004", and that adding "@inherits diameter_rfc4004" simply moves some code into different order (expected depending on how dependencies are placed in the dep tree).

So please ignore my previous comment, the PR seems to be working!

**Comment by Mikaka27** [comment]
> [@Mikaka27](https://github.com/Mikaka27) took me a while to fix our .dia dependencies in osmo-epdg when enabling indirect_inherits ;)
> 
> I confirm now code is generated even if removing "[@inherits](https://github.com/inherits) diameter_rfc4004", and that adding "[@inherits](https://github.com/inherits) diameter_rfc4004" simply moves some code into different order (expected depending on how dependencies are placed in the dep tree).
> 
> So please ignore my previous comment, the PR seems to be working!

Hi, we saw the potential for avp name conflicts and in turn breaking of existing applications, that's why we have this under a new option. Is that what you had to fix in your application, or is there something else?

**Comment by pespin** [comment]
> Hi, we saw the potential for avp name conflicts and in turn breaking of existing applications, that's why we have this under a new option. Is that what you had to fix in your application, or is there something else?

@Mikaka27 yes that was exactly it. See commits from 7f5d26f14b to b1b6237452 in https://gitea.osmocom.org/erlang/osmo-epdg/commits/branch/master

**Comment by Mikaka27** [comment]
Fixed by #10149
